### PR TITLE
Allow detached plugin location to be overridden

### PR DESCRIPTION
**Pull request by car-roll**
Alternative to #7301 

This PR allows subclasses of PluginManager to override the location where Jenkins will look to load the detached plugins from.

### Testing done

Overrode `getDetachedLocation` in a custom class that extended `PluginManager`. The new detached location pointed to `WEB-INF/plugins` instead. This allowed the detached plugins to successfully load even when it contained a plugin dependency that had not been detached.  The Jenkins instance loaded successfully and was operable. ATH successfully passed.
For unmodified Jenkins, ATH passed and I was able to start and operate the jenkins instance as expected.
Upgraded test:
 Installed a version of jenkins (2.346) that used an older version of sshd (3.237.v883d165a_c1d3) using `sshd-core`. Upgraded to jenkins 2.375 so that `sshd` was at 3.249.v2dc2ea_416e33 and was  now using `mina-sshd-api` plugins. The Jenkins instance started and operated as expected. I confirmed that the `mina-sshd-api' plugins were now installed after the upgrade (also confirmed they were not installed before upgrade)

### Proposed changelog entries

* Developer: Allow detached plugin location to be overridden.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7303"></a>

### CODE REVIEW

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/PluginManager.java
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java:634]:
```suggestion
     * Defines the location of the detached plugins in the WAR.
     * @return by default, {@code /WEB-INF/detached-plugins}
     * @since TODO
```

**Review by jglick — APPROVED** [review]
Again, seems reasonable, assuming it is proven to work in an end-to-end PR (presumably one only visible to @cloudbees, unfortunately).

**Review by rsandell — APPROVED** [review]

**Review by jtnord — APPROVED** [review]

**Comment by jglick** [comment]
Not yet demonstrated to work AFAIK.

**Review by basil — COMMENTED** [review]
The **Testing done** section is mandatory, not optional. Please fill it in.

**Comment by car-roll** [comment]
Sorry, switching back to draft. DId not mean to give  people the impression I was ready to merge. Still testing  other repos

**Review by basil — APPROVED** [review]
**Testing done** section looks reasonable. :+1:

**Comment by jglick** [comment]
> This allowed the detached plugins to successfully load

Can you expand on that a bit? You upgraded from an installation containing a release of some plugin with an old baseline which forced one or more detached plugins to be installed?

**Comment by car-roll** [comment]
ci tests look like they're being terminated early
ATH also has a failure `No such DSL method 'readYaml' found among steps` which I'm pretty sure is not related to the changes in this PR

**Review by NotMyFault — APPROVED** [review]
This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
